src/transformers/models/univnet/feature_extraction_univnet.py

```diff
@@ -0,0 +1,386 @@
+                "The argument `frame_signal_scale` is deprecated and will be removed in version 4.30.0 of Transformers",
+                FutureWarning,
+            )
+        if reduction_factor != 2.0:
+            warnings.warn(
+                "The argument `reduction_factor` is deprecated and will be removed in version 4.30.0 of Transformers",
+                FutureWarning,
+            )
+
+    # Based on tacotron2.audio_processing.dynamic_range_compression
+    # https://github.com/NVIDIA/tacotron2/blob/master/audio_processing.py#L78
+    def dynamic_range_compression(self, waveform: np.ndarray):
+        return np.log(np.clip(waveform, a_min=self.compression_clip_val, a_max=None) * self.compression_factor)
+
+    # Based on tacotron2.audio_processing.dynamic_range_compression
+    # https://github.com/NVIDIA/tacotron2/blob/master/audio_processing.py#L87
+    def dynamic_range_decompression(self, waveform: np.ndarray):
+        return np.exp(waveform) / self.compression_factor
+
+    def normalize(self, spectrogram):
+        return 2 * ((spectrogram - self.normalize_min) / (self.normalize_max - self.normalize_min)) - 1
+
+    def denormalize(self, spectrogram):
+        return self.normalize_min + (self.normalize_max - self.normalize_min) * ((spectrogram + 1) / 2)
+
+    def mel_spectrogram(self, waveform: np.ndarray) -> np.ndarray:
+        """
+        Calculates log MEL spectrograms from a batch of waveforms.
+
+        Args:
+            waveform (`np.ndarray` of shape `(length,)`):
+                The input waveform. This must be a single real-valued, mono waveform.
+
+        Returns:
+            `np.ndarray` containing a spectrogram of shape `(length, num_mel_bins)`.
+        """
+        # Do custom padding
+        # See https://github.com/maum-ai/univnet/blob/master/utils/stft.py#L84
+        waveform = np.pad(
+            waveform,
+            (int((self.n_fft - self.hop_length) / 2), int((self.n_fft - self.hop_length) / 2)),
+            mode="reflect",
+        )
+
+        # Get the complex spectrogram.
+        # Note: waveform must be unbatched currently due to the implementation of spectrogram(...).
+        complex_spectrogram = spectrogram(
+            waveform,
+            window=self.window,
+            frame_length=self.n_fft,
+            hop_length=self.hop_length,
+            fft_length=self.n_fft,
+            power=None,
+            center=self.center,
+            mel_filters=None,
+            mel_floor=None,
+        )
+
+        # Apply the MEL filter bank and MEL floor manually.
+        amplitude_spectrogram = np.sqrt(
+            np.real(complex_spectrogram) ** 2 + np.imag(complex_spectrogram) ** 2 + self.mel_floor
+        )
+        amplitude_spectrogram = np.matmul(self.mel_filters.T, amplitude_spectrogram)
+
+        # Perform spectral normalization to get the log mel spectrogram.
+        log_mel_spectrogram = self.dynamic_range_compression(amplitude_spectrogram)
```

Comment:
I tried supplying `mel_filters` and `mel_floor` directly to `spectrogram`, but I couldn't get the tests to pass because UnivNet uses a slightly different implementation to that of `spectrogram` when applying the MEL filter bank to the spectrogram. The `spectrogram` implementation does

https://github.com/huggingface/transformers/blob/feec56959afe480e57b2acc177111ae18a5ea757/src/transformers/audio_utils.py#L432-L433

whereas the UnivNet reference implementation ([original code](https://github.com/maum-ai/univnet/blob/9bb2b54838bb6d7ce767131cc7b8b61198bc7558/utils/stft.py#L92-L94)) first applies the equivalent of `mel_floor` while calculating the amplitude spectrogram:

```python
# `spectrogram` is a complex spectrogram of shape (*, N, T, 2)
spectrogram = torch.sqrt(spectrogram.pow(2).sum(-1) + (mel_floor))
```

and then applies the MEL filter bank without any floor:

```python
spectrogram = torch.matmul(mel_filters, spectrogram)
```

which I believe in general is not equivalent to the `spectrogram` implementation.